File: IA/pdf-a-md-regulacion.ts
// scripts/pdf-a-md-regulacion.ts
//
// Convierte un PDF de un código/ley de Guatemala al formato .md
// (front-matter + '## ARTICULO N...') que espera el pipeline de ingesta
// (segmentar.ts) que ya armamos para AlbaIA.
//
// npm install pdf-parse
//
// Uso:
//   npx tsx scripts/pdf-a-md-regulacion.ts entrada.pdf salida.md \
//     --ley "Codigo de Trabajo" --decreto "1441" --tema "laboral"

import fs from "fs";
import pdfParse from "pdf-parse";

interface Metadata {
  ley: string;
  decreto: string;
  tema: string;
  vigente_desde?: string;
  fuente?: string;
}

interface Articulo {
  titulo: string;
  texto: string;
}

async function extraerTexto(pdfPath: string): Promise<string> {
  const buffer = fs.readFileSync(pdfPath);
  const data = await pdfParse(buffer);
  // pdf-parse a veces deja caracteres de control invisibles (saltos de forma,
  // bytes de control de fuentes, etc.) que ensucian el markdown resultante.
  return data.text.replace(/[\x00-\x08\x0B\x0E-\x1F]/g, "").replace(/\r/g, "");
}

/**
 * Elimina líneas que se repiten en un porcentaje alto de "apariciones"
 * (encabezados/pies de página tipo 'OFICINA NACIONAL DE SERVICIO CIVIL'
 * o 'El documento fue generado para el uso exclusivo de...').
 */
function quitarEncabezadosRepetidos(texto: string, umbral = 0.5): string {
  const lineas = texto.split("\n");
  const conteo = new Map<string, number>();

  for (const linea of lineas) {
    const key = linea.trim();
    if (!key) continue;
    conteo.set(key, (conteo.get(key) ?? 0) + 1);
  }

  // \f = salto de página (form feed) que suele dejar pdf-parse entre páginas
  const totalPaginasAprox = (texto.match(/\f/g)?.length ?? 0) + 1;
  const umbralRepeticiones = Math.max(3, totalPaginasAprox * umbral);

  const lineasAQuitar = new Set(
    [...conteo.entries()]
      .filter(([linea, n]) => n >= umbralRepeticiones && linea.length > 3)
      .map(([linea]) => linea)
  );

  return lineas.filter((l) => !lineasAQuitar.has(l.trim())).join("\n");
}

/**
 * Divide el texto en artículos usando el patrón 'ARTICULO N' / 'ARTÍCULO N'
 * (con o sin tilde, mayúsculas o minúsculas) al inicio de línea.
 */
function segmentarArticulos(texto: string): Articulo[] {
  const patron = /^\s*(art[íi]culo\s+\d+[.\-]?.*)$/gim;

  const posiciones: number[] = [];
  const titulos: string[] = [];
  let match: RegExpExecArray | null;

  while ((match = patron.exec(texto)) !== null) {
    posiciones.push(match.index);
    titulos.push(match[1].trim());
  }

  if (posiciones.length === 0) return [];

  const articulos: Articulo[] = [];
  for (let i = 0; i < posiciones.length; i++) {
    const fin = i + 1 < posiciones.length ? posiciones[i + 1] : texto.length;
    let bloque = texto.slice(posiciones[i], fin).trim();
    bloque = bloque.replace(/\n{3,}/g, "\n\n");
    bloque = bloque.replace(/[ \t]{2,}/g, " ");
    articulos.push({ titulo: titulos[i], texto: bloque });
  }

  return articulos;
}

function construirMarkdown(articulos: Articulo[], metadata: Metadata): string {
  const frontMatter = ["---"];
  for (const [k, v] of Object.entries(metadata)) {
    if (v) frontMatter.push(`${k}: "${v}"`);
  }
  frontMatter.push("---\n");

  const cuerpo = articulos.map((art) => {
    const [primeraLinea, ...resto] = art.texto.split("\n");
    return `## ${primeraLinea.trim()}\n\n${resto.join("\n").trim()}\n`;
  });

  return frontMatter.join("\n") + "\n" + cuerpo.join("\n");
}

function parseArgs(argv: string[]) {
  const [pdfEntrada, mdSalida, ...rest] = argv;
  const flags: Record<string, string> = {};
  for (let i = 0; i < rest.length; i += 2) {
    flags[rest[i].replace(/^--/, "")] = rest[i + 1];
  }
  return { pdfEntrada, mdSalida, flags };
}

async function main() {
  const { pdfEntrada, mdSalida, flags } = parseArgs(process.argv.slice(2));

  if (!pdfEntrada || !mdSalida || !flags.ley || !flags.decreto || !flags.tema) {
    console.error(
      "Uso: pdf-a-md-regulacion.ts entrada.pdf salida.md --ley '...' --decreto '...' --tema '...' [--vigente_desde '...'] [--fuente '...']"
    );
    process.exit(1);
  }

  let texto = await extraerTexto(pdfEntrada);
  texto = quitarEncabezadosRepetidos(texto);
  const articulos = segmentarArticulos(texto);

  console.log(`Artículos detectados: ${articulos.length}`);

  const markdown = construirMarkdown(articulos, {
    ley: flags.ley,
    decreto: flags.decreto,
    tema: flags.tema,
    vigente_desde: flags.vigente_desde,
    fuente: flags.fuente,
  });

  fs.writeFileSync(mdSalida, markdown, "utf-8");
  console.log(`Guardado en ${mdSalida}`);
}

main();

File: IA/pdf_a_md_regulacion.py
"""
Convierte un PDF de un código/ley de Guatemala al formato .md
(front-matter + '## ARTICULO N...') que espera el pipeline de ingesta
(segmentar.ts / ingest_regulaciones.py) que ya armamos para AlbaIA.

Uso:
    python3 pdf_a_md_regulacion.py entrada.pdf salida.md \
        --ley "Codigo de Trabajo" --decreto "1441" --tema "laboral"

Requiere: pdftotext (poppler-utils, ya viene instalado en este entorno)
"""

import argparse
import re
import subprocess
from collections import Counter


def extraer_texto(pdf_path: str) -> str:
    resultado = subprocess.run(
        ["pdftotext", "-layout", pdf_path, "-"],
        capture_output=True, text=True, check=True
    )
    return resultado.stdout


def quitar_encabezados_repetidos(texto: str, umbral: float = 0.5) -> str:
    """
    Elimina líneas que se repiten en un porcentaje alto de "apariciones"
    (encabezados/pies de página tipo 'OFICINA NACIONAL DE SERVICIO CIVIL'
    o 'El documento fue generado para el uso exclusivo de...').
    """
    lineas = texto.split("\n")
    conteo = Counter(l.strip() for l in lineas if l.strip())
    total_paginas_aprox = texto.count("\x0c") + 1  # form feed = salto de página
    umbral_repeticiones = max(3, total_paginas_aprox * umbral)

    lineas_a_quitar = {
        l for l, n in conteo.items()
        if n >= umbral_repeticiones and len(l) > 3
    }

    limpio = [l for l in lineas if l.strip() not in lineas_a_quitar]
    return "\n".join(limpio)


def segmentar_articulos(texto: str) -> list[dict]:
    """
    Divide el texto en artículos usando el patrón 'ARTICULO N' / 'ARTÍCULO N'
    (con o sin tilde, mayúsculas o minúsculas) al inicio de línea.
    """
    patron = re.compile(r"(?im)^\s*(art[íi]culo\s+\d+[\.\-]?.*)$")

    posiciones = [m.start() for m in patron.finditer(texto)]
    titulos = [m.group(1).strip() for m in patron.finditer(texto)]

    if not posiciones:
        return []

    articulos = []
    for i, pos in enumerate(posiciones):
        fin = posiciones[i + 1] if i + 1 < len(posiciones) else len(texto)
        bloque = texto[pos:fin].strip()
        # Colapsar múltiples líneas en blanco/espacios de más
        bloque = re.sub(r"\n{3,}", "\n\n", bloque)
        bloque = re.sub(r"[ \t]{2,}", " ", bloque)
        articulos.append({"titulo": titulos[i], "texto": bloque})

    return articulos


def construir_markdown(articulos: list[dict], metadata: dict) -> str:
    front_matter_lineas = ["---"]
    for k, v in metadata.items():
        front_matter_lineas.append(f'{k}: "{v}"')
    front_matter_lineas.append("---\n")

    cuerpo = []
    for art in articulos:
        # Aseguramos que cada artículo quede como header '## '
        primera_linea, _, resto = art["texto"].partition("\n")
        cuerpo.append(f"## {primera_linea.strip()}\n\n{resto.strip()}\n")

    return "\n".join(front_matter_lineas) + "\n".join(cuerpo)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("pdf_entrada")
    parser.add_argument("md_salida")
    parser.add_argument("--ley", required=True)
    parser.add_argument("--decreto", required=True)
    parser.add_argument("--tema", required=True)
    parser.add_argument("--vigente_desde", default="")
    parser.add_argument("--fuente", default="")
    args = parser.parse_args()

    texto = extraer_texto(args.pdf_entrada)
    texto = quitar_encabezados_repetidos(texto)
    articulos = segmentar_articulos(texto)

    print(f"Artículos detectados: {len(articulos)}")

    metadata = {
        "ley": args.ley,
        "decreto": args.decreto,
        "tema": args.tema,
        "vigente_desde": args.vigente_desde,
        "fuente": args.fuente,
    }

    markdown = construir_markdown(articulos, metadata)

    with open(args.md_salida, "w", encoding="utf-8") as f:
        f.write(markdown)

    print(f"Guardado en {args.md_salida}")


if __name__ == "__main__":
    main()
